server/src/main/java/org/opensearch/index/compositeindex/CompositeIndexMetadata.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.compositeindex;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.CorruptIndexException;
+import org.apache.lucene.store.IndexInput;
+import org.opensearch.index.compositeindex.datacube.startree.meta.StarTreeMetadata;
+import org.opensearch.index.mapper.CompositeMappedFieldType;
+
+import java.io.IOException;
+
+import static org.opensearch.index.compositeindex.CompositeIndexConstants.VERSION;
+
+/**
+ * This class represents the metadata of a Composite Index, which includes information about
+ * the composite field name, type, and the specific metadata for the type of composite field
+ * (e.g., Tree metadata).
+ *
+ * @opensearch.experimental
+ */
+public class CompositeIndexMetadata {
+
+    private static final Logger logger = LogManager.getLogger(CompositeIndexMetadata.class);
+    private final String compositeFieldName;
+    private final CompositeMappedFieldType.CompositeFieldType compositeFieldType;
+    private final StarTreeMetadata starTreeMetadata;
```

Comment:
Lets use inheritance for any implementation specific metadata information. Composition will result in each new type getting added to CompositeIndexMetadata and break the compatibility.

CompositeIndexMetadata should have high level metadata and starTreeFieldMeta should have low level metadata.